graphbolt/src/cnumpy.cc

```diff
@@ -0,0 +1,192 @@
+/**
+ *  Copyright (c) 2023 by Contributors
+ * @file cnumpy.cc
+ * @brief Numpy File Fetecher class.
+ */
+
+#include "cnumpy.h"
+
+#include <fcntl.h>
+#include <liburing.h>
+#include <omp.h>
+#include <stdint.h>
+#include <stdlib.h>
+
+#include <cstdlib>
+#include <cstring>
+#include <fstream>
+#include <regex>
+#include <stdexcept>
+
+#define ALIGNMENT 4096
+
+namespace graphbolt {
+namespace storage {
+
+/**
+ * @brief Read disk numpy file based on given index and transform to tensor.
+ */
+torch::Tensor OnDiskNpyArray::index_select_iouring(torch::Tensor idx) {
+  int feature_fd = open(filename.c_str(), O_RDONLY | O_DIRECT);
+
+  int64_t feature_size = feat_dim * word_size;
+  int64_t num_idx = idx.numel();
+
+  // Multi-thread number.
+  int numthd = 4;
+  omp_set_num_threads(numthd);
+  // Each group size for reader.
+  int64_t group_size = 512;
+
+  char *read_buffer =
+      (char *)aligned_alloc(ALIGNMENT, ALIGNMENT * 2 * group_size * numthd);
+  char *result_buffer =
+      (char *)aligned_alloc(ALIGNMENT, feature_size * num_idx);
+
+  auto idx_data = idx.data_ptr<int64_t>();
+
+  // Init io_uring fetcher.
+  struct io_uring ring[numthd];
+  for (int64_t t = 0; t < numthd; t++) {
+    io_uring_queue_init(512, &ring[t], 0);
+  }
+
+  // Record the inside offsets of feteched features.
+  int64_t residual[group_size * numthd];
+
+#pragma omp parallel for num_threads(numthd)
```

Comment:
Why not use torch parallel?

Reply:
I have not used it yet, can you show me an example?

Reply:
If we use torch parallel, we need a way to limit the number of threads used later for performance. We will want to have the rest of the threads running other ops.

Reply:
Example: https://github.com/dmlc/dgl/blob/1ec0092e1598fdc43ccc5baf22872386d7ad865e/graphbolt/src/fused_csc_sampling_graph.cc#L306-L322

Reply:
https://github.com/pytorch/pytorch/blob/5cc511f72fe073bbd8c10d796d72dce67f5cd5c4/aten/src/ATen/ParallelFuture.h#L10

We might want to use this to make the disk read asyncronous so that we can overlap it with other computations.

Reply:
Sure, for the asynchronous execution we may optimize it in the future.